src/mapper/pkg/otelexporter/otel_exporter.go

```diff
@@ -0,0 +1,140 @@
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+)
+
+type OtelExporter struct {
+	config        Config
+	counter       metric.Int64Counter
+	intentsHolder *intentsstore.IntentsHolder
+}
+
+func newResource() (*resource.Resource, error) {
+	return resource.Merge(resource.Default(),
+		resource.NewWithAttributes(semconv.SchemaURL,
+			semconv.OTelLibraryName("otterize/network-mapper"),
+			semconv.OTelLibraryVersion(version.Version()),
+		))
+}
+
+const DefaultMetricEndpoint = "ingest.lightstep.com:443"
+
+// uses same name as expected in opentelemetry-collector-contrib's servicegraphprocessor
+const CounterMetricName = "traces_service_graph_request_total"
+const ClientAttributeName = "client"
+const ServerAttributeName = "server"
+
+func getenv(key, fallback string) string {
+	value := os.Getenv(key)
+	if len(value) == 0 {
+		return fallback
+	}
+	return value
+}
+
+func newMeterProvider(ctx context.Context, res *resource.Resource, exportInterval time.Duration) (*sdk.MeterProvider, error) {
+	metricExporter, err := otlpmetricgrpc.New(ctx,
+		otlpmetricgrpc.WithEndpoint(getenv("OTEL_EXPORTER_OTLP_ENDPOINT", DefaultMetricEndpoint)),
+		otlpmetricgrpc.WithHeaders(map[string]string{
+			"lightstep-access-token": os.Getenv("LS_ACCESS_TOKEN"),
+		}),
+		otlpmetricgrpc.WithTimeout(7*time.Second),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if viper.GetBool(sharedconfig.DebugKey) {
+		stdOutExporter, err := stdoutmetric.New()
+		if err != nil {
+			return nil, err
+		}
+		meterProvider := sdk.NewMeterProvider(
+			sdk.WithResource(res),
+			sdk.WithReader(sdk.NewPeriodicReader(stdOutExporter,
+				sdk.WithInterval(1*time.Second))),
+			sdk.WithReader(sdk.NewPeriodicReader(metricExporter,
+				sdk.WithInterval(exportInterval))),
+		)
+		return meterProvider, nil
+	}
+
+	meterProvider := sdk.NewMeterProvider(
+		sdk.WithResource(res),
+		sdk.WithReader(sdk.NewPeriodicReader(metricExporter,
+			sdk.WithInterval(exportInterval))),
+	)
+	return meterProvider, nil
+}
+
+func NewOtelExporter(ctx context.Context, ih *intentsstore.IntentsHolder, config Config) *OtelExporter {
+	res, err := newResource()
+	if err != nil {
+		panic(err)
+	}
+
+	meterProvider, err := newMeterProvider(ctx, res, config.ExportInterval)
+	if err != nil {
+		panic(err)
+	}
+
+	// TODO: this is not the right place to handle shutdown
+	// defer func() {
+	// 	err := meterProvider.Shutdown(context.Background())
+	// 	if err != nil {
+	// 		logrus.Fatalln(err)
+	// 	}
+	// }()
+
+	var meter = meterProvider.Meter("otelexporter")
+	edgeCounter, err := meter.Int64Counter(
+		CounterMetricName,
+		metric.WithDescription("Count of edges between two nodes"),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return &OtelExporter{
+		intentsHolder: ih,
+		config:        config,
+		counter:       edgeCounter,
+	}
+}
+
+func (o *OtelExporter) countDiscoveredIntents(ctx context.Context) {
+	for _, intent := range o.intentsHolder.GetNewIntentsSinceLastGet() {
```

Comment:
This function `GetNewIntentsSinceLastGet` is used by the network mapper to export intents to Otterize Cloud, and it gets reset whenever it's read. So if both OtelExporter and Otterize Cloud are active at the same time, they will interfere with each other, and they will not both see all the data.

To deal with this, please add a third object that calls `GetNewIntentsSinceLastGet` periodically and then calls both `CloudUploader` and `OtelExporter` with the new intents. Or you can change `IntentsHolder` so that it can have callbacks registered to it and calls out to the callbacks (if any, since both Otterize Cloud and OtelExporter can be disabled) instead of waiting for them to query its copy and clearing it. They can keep those intents in a slice internally and then just send them when ready, without calling out to `IntentsHolder`.

Reply:
Thanks, now updated with a callback approach.